File: physical/src/main/java/vax/openglue/Uniform.cs

using System.Collections.Generic;

using OpenTK;
using OpenTK.Graphics.OpenGL;

using physical.math;

namespace physical {
    public interface IUniform {
        void updateGl ( int uniformLocation );

        string getName ();

        string getPrefix ();
    }

    public abstract class Uniform<T> : IUniform {
        protected T value;
        protected string name;

        public Uniform ( string name, T value ) {
            this.value = value;
            this.name = name;
        }

        public string getName () {
            return name;
        }

        override public string ToString () {
            return getPrefix() + " " + getName() + "[" + value + "]";
        }

        public string ToShaderString () {
            return "uniform " + getPrefix() + " " + getName() + ";\n";
        }

        abstract public void updateGl ( int uniformLocation );

        abstract public string getPrefix ();
    }

    public class UniformManager {
        public static readonly List<IUniform> uniforms = new List<IUniform>();
        public static readonly Dictionary<IUniform,int> uniformMap = new Dictionary<IUniform,int>();
        public static readonly Dictionary<string,IUniform> uniformNameMap = new Dictionary<string,IUniform>();

        public void addUniform ( IUniform iUniform ) {
            uniforms.Add( iUniform );
        }

        public void addUniforms ( params IUniform[] iUniforms ) {
            foreach ( IUniform iUniform in iUniforms )
                uniforms.Add( iUniform );
        }

        public IUniform getIUniform ( string name ) {
            return uniformNameMap[name];
        }

        public int getLocation ( IUniform iUniform ) {
            return uniformMap[iUniform];
        }

        public void init ( int shaderProgramHandle ) {
            foreach ( IUniform uniform in uniforms ) {
                int uniformLocation = GL.GetUniformLocation( shaderProgramHandle, uniform.getName() );
                uniformMap.Add( uniform, uniformLocation );
                uniformNameMap.Add( uniform.getName(), uniform );
                uniform.updateGl( uniformLocation );
            }
        }

        public void updateGl ( string uniformName ) {
            updateGl( getIUniform( uniformName ) );
        }

        public void updateGl ( IUniform iUniform ) {
            iUniform.updateGl( getLocation( iUniform ) );
        }

        public void updateGl () {
            foreach ( IUniform uniform in uniforms )
                updateGl( uniform );
        }
    }

    /*
    public class Uniform1f : Uniform<float[]> {
        public Uniform1f ( String name, ref float[] value ) : base( name, ref value ) {
        }

        override public void updateGl ( int uniformLocation ) {
            GL.Uniform1( uniformLocation, value[0] );
        }
    }


    public class Uniform2f : Uniform<Vector2> {
        public Uniform2f ( String name, ref Vector2 value ) : base( name, ref value ) {
        }

        override public void updateGl ( int uniformLocation ) {
            GL.Uniform2( uniformLocation, ref value );
        }
    }


    public class Uniform3f : Uniform<Vector3> {
        public Uniform3f ( String name, ref Vector3 value ) : base( name, ref value ) {
        }

        override public void updateGl ( int uniformLocation ) {
            GL.Uniform3( uniformLocation, ref value );
        }
    }

    public class Uniform4f : Uniform<Vector4> {
        public Uniform4f ( String name, ref Vector4 value ) : base( name, ref value ) {
        }

        override public void updateGl ( int uniformLocation ) {
            GL.Uniform4( uniformLocation, ref value );
        }
    }

    public class UniformMatrix4 : Uniform<Matrix4> {
        public UniformMatrix4 ( String name, Matrix4 value ) : base( name, ref value ) {
        }

        override public void updateGl ( int uniformLocation ) {
            GL.UniformMatrix4( uniformLocation, false, value );
        }
    }
    */

    public class Uniform1f : Uniform<Value1f> {
        public Uniform1f ( string name, Value1f value ) : base( name, value ) {
        }

        override public void updateGl ( int uniformLocation ) {
            GL.Uniform1( uniformLocation, value.Value );
        }

        override public string getPrefix () {
            return "float";
        }
    }

    public class Uniform2f : Uniform<Vector2f> {
        public Uniform2f ( string name, Vector2f value ) : base( name, value ) {
        }

        override public void updateGl ( int uniformLocation ) {
            GL.Uniform2( uniformLocation, 1, value.Data );
        }

        override public string getPrefix () {
            return "vec2";
        }
    }


    public class Uniform3f : Uniform<Vector3f> {
        public Uniform3f ( string name, Vector3f value ) : base( name, value ) {
        }

        override public void updateGl ( int uniformLocation ) {
            GL.Uniform3( uniformLocation, 1, value.Data );
        }

        override  public string getPrefix () {
            return "vec3";
        }
    }


    public class Uniform4f : Uniform<Vector4f> {
        public Uniform4f ( string name, Vector4f value ) : base( name, value ) {
        }

        override public void updateGl ( int uniformLocation ) {
            GL.Uniform4( uniformLocation, 1, value.Data );
        }

        override  public string getPrefix () {
            return "vec4";
        }
    }

    public class UniformMatrix4f : Uniform<Matrix4f> {
        public UniformMatrix4f ( string name, Matrix4f value ) : base( name, value ) {
        }

        override public void updateGl ( int uniformLocation ) {
            GL.UniformMatrix4( uniformLocation, 1, false, value.Data );
        }

        override public string getPrefix () {
            return "mat4";
        }
    }
}

File: physical/src/main/java/vax/openglue/Uniform.java
package vax.openglue;

import java.lang.reflect.Field;
import vax.math.*;

/**

 @author toor
 */
public interface Uniform {
    void updateGL ( OpenGLUE gl, int uniformLocation );

    String getName ();

    String getPrefix ();

    String toShaderString ();

    static Uniform from ( Field field, Object value ) {
        try {
            return from( field.getName(), field.get( value ) );
        } catch (IllegalArgumentException | IllegalAccessException ex) {
            throw new RuntimeException( ex );
        }
    }

    static Uniform from ( String name, Object value ) {
        if ( value instanceof Value1i ) {
            return new U1i( name, (Value1i) value );
        } else if ( value instanceof Vector2i ) {
            return new U2i( name, (Vector2i) value );
        } else if ( value instanceof Value1f ) {
            return new U1f( name, (Value1f) value );
        } else if ( value instanceof Vector2f ) {
            return new U2f( name, (Vector2f) value );
        } else if ( value instanceof Vector3f ) {
            return new U3f( name, (Vector3f) value );
        } else if ( value instanceof Vector4f ) {
            return new U4f( name, (Vector4f) value );
        } else if ( value instanceof Matrix4f ) {
            return new UMatrix4f( name, (Matrix4f) value );
        }

        throw new IllegalArgumentException( ( ( value instanceof Uniform ) ? "uknown uniform type: '" : "not a Uniform type: '" )
                + value.getClass() + "'" );
    }

    static U1i from ( String name, Value1i value ) {
        return new U1i( name, value );
    }

    static U2i from ( String name, Vector2i value ) {
        return new U2i( name, value );
    }

    static U1f from ( String name, Value1f value ) {
        return new U1f( name, value );
    }

    static U2f from ( String name, Vector2f value ) {
        return new U2f( name, value );
    }

    static U3f from ( String name, Vector3f value ) {
        return new U3f( name, value );
    }

    static U4f from ( String name, Vector4f value ) {
        return new U4f( name, value );
    }

    static UMatrix4f from ( String name, Matrix4f value ) {
        return new UMatrix4f( name, value );
    }

    abstract class AbstractUniform<T> implements Uniform {
        protected T value;
        protected String name;

        public AbstractUniform ( String name, T value ) {
            this.value = value;
            this.name = name;
        }

        @Override
        public String getName () {
            return name;
        }

        @Override
        abstract public void updateGL ( OpenGLUE gl, int uniformLocation );

        @Override
        abstract public String getPrefix ();

        @Override
        public String toString () {
            return getPrefix() + " " + getName() + "( " + value + " )";
        }

        @Override
        public String toShaderString () {
            return "uniform " + getPrefix() + " " + getName() + ";\n";
        }
    }

    class U1i extends AbstractUniform<Value1i> {
        public U1i ( String name, Value1i value ) {
            super( name, value );
        }

        @Override
        public void updateGL ( OpenGLUE gl, int uniformLocation ) {
            gl.glUniform1i( uniformLocation, value.getValue() );
        }

        @Override
        public String getPrefix () {
            return "int";
        }
    }

    class U2i extends AbstractUniform<Vector2i> {
        public U2i ( String name, Vector2i value ) {
            super( name, value );
        }

        @Override
        public void updateGL ( OpenGLUE gl, int uniformLocation ) {
            gl.glUniform2i( uniformLocation, gl.ueGetGLUtils().wrap( value.getData() ) );
        }

        @Override
        public String getPrefix () {
            return "ivec2";
        }
    }

    class U1f extends AbstractUniform<Value1f> {
        public U1f ( String name, Value1f value ) {
            super( name, value );
        }

        @Override
        public void updateGL ( OpenGLUE gl, int uniformLocation ) {
            gl.glUniform1f( uniformLocation, gl.ueGetGLUtils().wrap( value.get() ) );
        }

        @Override
        public String getPrefix () {
            return "float";
        }
    }

    class U2f extends AbstractUniform<Vector2f> {
        public U2f ( String name, Vector2f value ) {
            super( name, value );
        }

        @Override
        public void updateGL ( OpenGLUE gl, int uniformLocation ) {
            gl.glUniform2f( uniformLocation, gl.ueGetGLUtils().wrap( value.getData() ) );
        }

        @Override
        public String getPrefix () {
            return "vec2";
        }
    }

    class U3f extends AbstractUniform<Vector3f> {
        public U3f ( String name, Vector3f value ) {
            super( name, value );
        }

        @Override
        public void updateGL ( OpenGLUE gl, int uniformLocation ) {
            gl.glUniform3f( uniformLocation, gl.ueGetGLUtils().wrap( value.getData() ) );
        }

        @Override
        public String getPrefix () {
            return "vec3";
        }
    }

    class U4f extends AbstractUniform<Vector4f> {
        public U4f ( String name, Vector4f value ) {
            super( name, value );
        }

        @Override
        public void updateGL ( OpenGLUE gl, int uniformLocation ) {
            gl.glUniform4f( uniformLocation, gl.ueGetGLUtils().wrap( value.getData() ) );
        }

        @Override
        public String getPrefix () {
            return "vec4";
        }
    }

    class UMatrix4f extends AbstractUniform<Matrix4f> {
        public UMatrix4f ( String name, Matrix4f value ) {
            super( name, value );
        }

        @Override
        public void updateGL ( OpenGLUE gl, int uniformLocation ) {
            gl.glUniformMatrix4f( uniformLocation, gl.ueGetGLUtils().wrap( value.getData() ) );
        }

        @Override
        public String getPrefix () {
            return "mat4";
        }
    }
}
